# EXAMEN - Convocatoria 1 - Programación
Resolver el siguientes problemas de regresión lineal y clasificación.

## 1. Problema de Regresión lineal:

### 1.1. Generación de los datos:

Mediante la función `make_regressión` de sklearn, consultar su como emplearla para generar un conjunto de 2000 datos, 10 características, un ruido de 1 y fijar el parámetro ramdom_state=42.

In [42]:
# Importación de librerías
import numpy as np
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, make_scorer
from sklearn.model_selection import GridSearchCV
import pandas as pd


In [14]:
# Genero el conjunto de datos

A , b = make_regression(
    n_samples=2000, # Número de datos / filas
    n_features=10, # Número de atributos / columnas
    noise=1,
    random_state=42 # semilla de aleatoriedad
)

print(f'Matriz A:')
display(A)
print()
print(f'Vector b:')
display (b)

Matriz A:


array([[-1.38301781, -0.18140603,  0.73752806, ...,  0.94782078,
        -2.09317802, -1.07681697],
       [-0.81819888,  0.77836108,  0.08840689, ..., -0.55118572,
        -0.00337446,  0.95530521],
       [ 0.0358026 , -0.22430683, -0.55512516, ..., -0.06945514,
         0.13814756,  0.73221807],
       ...,
       [ 0.67683765,  0.94916607,  1.43506541, ..., -1.21176821,
        -1.08110935,  1.14145458],
       [ 1.91268813, -0.3217527 ,  0.57114124, ..., -0.03248104,
        -1.32484578, -0.98463369],
       [-0.83710633, -0.53516427,  1.3573415 , ...,  1.69045267,
         0.2834766 , -1.42886615]], shape=(2000, 10))


Vector b:


array([-142.12531758,   -9.39375045, -184.25199351, ...,  102.28031061,
        104.18531736,  102.86648815], shape=(2000,))

### 1.2. Partición de datos externa.
Realizar una partición externa de tipo hold-out seleccionando un 20% de los datos para test (fijar una semilla en 42).

In [17]:

# Hago la partición externa hold-out (80% entrenamiento, 20% test)
A_train, A_test, b_train, b_test = train_test_split(
    A, b,
    test_size=0.20, # Se crea el 20% de datos para el test
    random_state=42 # Se fija de nuevo la semilla
)

print(f"Tamaño A_train:")
display(A_train.shape)
print(f"Tamaño de A_test:")
display(A_test.shape)

Tamaño A_train:


(1600, 10)

Tamaño de A_test:


(400, 10)

### 1.3. Estandarización de los datos de train y test.
Utilizar el método StandardScaler().

In [29]:
escalado = StandardScaler()

# Pasamos a ajustar y transformar los datos de entrenamiento
A_train_std = escalado.fit_transform(A_train)

# Transformo los datos de testeo
A_test_std = escalado.transform(A_test)

# Hacemos el ajuste solo en los datos de entrenamiento,
# de lo contrario, si lo hiciéramos con los datos de testeo
# estaríamos haciendo "fuga de datos"

print(f'Datos estandarizados de entrenamiento:\n')
display(A_train_std)
print()
print(f'Datos estandarizados de testeo:\n')
display(A_test_std)

Datos estandarizados de entrenamiento:



array([[-1.19076762, -0.01134408, -1.55877614, ..., -0.23740193,
        -0.04082145,  0.24115295],
       [ 0.96979971, -1.71427615, -0.08050323, ...,  0.21801361,
        -0.22351523, -0.06411872],
       [-0.79960803,  0.07910062,  1.98094995, ...,  0.375499  ,
         1.2893963 , -0.95969181],
       ...,
       [ 0.17819919, -1.06192013,  1.6839707 , ..., -2.22385367,
         0.80499388, -0.08001885],
       [-0.26367917,  0.53537958,  0.4616912 , ...,  0.79674382,
         1.04704896,  0.98481908],
       [-0.13958716, -0.86939834,  0.99460438, ...,  0.67928441,
        -0.38547275, -1.29668648]], shape=(1600, 10))


Datos estandarizados de testeo:



array([[ 2.05593161, -1.00209523,  0.43589986, ...,  1.61921505,
         0.40009269,  0.04268114],
       [-0.82042042, -1.75173872, -0.86062949, ...,  0.43798797,
        -0.62569527,  0.02536138],
       [ 0.52227587, -1.23389989, -1.74149603, ..., -0.71197025,
         2.11887195,  1.92017821],
       ...,
       [-1.28566264, -0.35813671,  1.57531421, ..., -0.23828573,
        -0.84099637, -0.89773796],
       [ 0.22018383,  0.55797362, -0.44580661, ..., -0.20035825,
        -0.39735401, -0.88286677],
       [-0.74538327,  0.27002537,  2.14752248, ..., -0.87384334,
        -0.94239601, -0.71455306]], shape=(400, 10))

### 1.4. Creación de modelos e hiperparametrización mediante el método de sklearn GridSearchCV:

Instrucciones:

- Cree los siguientes modelos de regresión en un diccionario:
  - Regresión Lineal (OLS).
  - Regresión Lasso
  - Regresión Rígida
  - Vecinos más cercanos (KNN)

- Cree las siguiente métricas en un diccionario para evaluar:
  - MAE
  - MSE
  - RMSE
  - R2

- Haga una búsqueda de los siguientes hiperparámetros mediante GridSearchCV:

```
# Hiperparametros
parameters = {'OLS':{},
              'Lasso':{'alpha':(0.1,0.5,1,5,10,50,100)},
              'Ridge':{'alpha':(0.1,0.5,1,5,10,50,100)},
              'KNN':{'n_neighbors':np.arange(1,15),
                     'weights':('uniform','distance'),
                     'metric':('euclidean','manhattan','cosine')
                     }
              }
```

In [37]:
# Diccionario de modelos
modelos = {
    'Regresión Lineal (OLS)': LinearRegression(),
    'Regresión Lasso': Lasso(),
    'Regresión Rígida': Ridge(),
    'Vecinos más cercanos (KNN)': KNeighborsRegressor()
}

# Diccionario de métricas
metricas = {
    'MAE': 'neg_mean_absolute_error',
    'MSE': 'neg_mean_squared_error',
    'RMSE': 'neg_root_mean_squared_error',
    'R2': 'r2'
}

# Hiperparámetros

parametros = {
    'Regresión Lineal (OLS)': {},
    'Regresión Lasso': {'alpha': (0.1, 0.5, 1, 5, 10, 50, 100)},
    'Regresión Rígida': {'alpha': (0.1, 0.5, 1, 5, 10, 50, 100)},
    'Vecinos más cercanos (KNN)': {
        'n_neighbors': np.arange(1, 15),
        'weights': ('uniform', 'distance'),
        'metric': ('euclidean', 'manhattan', 'cosine')
    }
}


In [41]:
# Se crea un diccionario para almacenar los resultados finales
resultados_grid = {}

# Itero cada modelo del diccionario que creamos antes
for nombre in modelos:
    print(f"Búsqueda iniciada para: {nombre}...")
    
    # Creanis el buscador usando GridSearchCV
    grid_search = GridSearchCV(
        estimator=modelos[nombre], # Elegimos el modelo que creamos antes
        param_grid=parametros[nombre], # Elegimos los hiperparámetros de antes.
        scoring=metricas, # Usando el diccionario de métricas creado antes.
        refit='MAE', # Paara buscar el mejor basado en el MAE
        cv=5, # Validación cruzada de 5 "folds"
        n_jobs=-1 # Usar todos los núcleos para ir más rápido
    )
    
    # Entrenamos con los datos de entrenamiento estandarizados
    grid_search.fit(A_train_std, b_train)
    
    # Guardamos el objeto entrenado en el diccionario de resultados
    resultados_grid[nombre] = grid_search
    
    # Mostramos un pequeño resumen por pantalla
    print(f"---Mejores parámetros obtenidos: {grid_search.best_params_}")
    print(f"---Mejor MAE: {-grid_search.best_score_:.4f}")
    print("-" * 60)

print("La búsqueda ha sido finalizada")

Búsqueda iniciada para: Regresión Lineal (OLS)...
---Mejores parámetros obtenidos: {}
---Mejor MAE: 0.8143
------------------------------------------------------------
Búsqueda iniciada para: Regresión Lasso...
---Mejores parámetros obtenidos: {'alpha': 0.1}
---Mejor MAE: 0.8599
------------------------------------------------------------
Búsqueda iniciada para: Regresión Rígida...
---Mejores parámetros obtenidos: {'alpha': 0.1}
---Mejor MAE: 0.8148
------------------------------------------------------------
Búsqueda iniciada para: Vecinos más cercanos (KNN)...
---Mejores parámetros obtenidos: {'metric': 'cosine', 'n_neighbors': np.int64(14), 'weights': 'distance'}
---Mejor MAE: 54.2786
------------------------------------------------------------
La búsqueda ha sido finalizada


### 1.5. Evaluación de los modelos sobre el conjunto de test.
- Entrenar los modelos anteriores utilizando todos los datos de entrenamiento y quédese con el mejor modelo de cada uno de los modelos de regresión arrojados por GridSearchCV. Es decir, debe tener un mejor modelo por cada regresión (OLS, Lasso, Rígida y KNN)
- Evaluar su rendimiento sobre el conjunto de test mostrando para cada uno de los modelos

In [43]:
# Creamos otro diccionario para almacenar las métricas finales del testeo
resumen_test = []

for nombre in resultados_grid:
    # Extraemos el mejor estimador del modelo ya entrenado con los mejores parámetros
    mejor_modelo = resultados_grid[nombre].best_estimator_
    
    # Hacemos la predicción sobre el conjunto de testeo
    b_pred = mejor_modelo.predict(A_test_std)
    
    # Calculamos las métricas de test
    mae = mean_absolute_error(b_test, b_pred)
    mse = mean_squared_error(b_test, b_pred)
    rmse = np.sqrt(mse)
    r2 = r2_score(b_test, b_pred)
    
    # Almaceno los resultados en una lista
    resumen_test.append({
        'Modelo': nombre,
        'MAE Test': mae,
        'MSE Test': mse,
        'RMSE Test': rmse,
        'R2 Test': r2
    })

# Convertimos a DataFrame para mejorar la visualzación
df_resultados = pd.DataFrame(resumen_test)
display(df_resultados)

,Modelo,MAE Test,MSE Test,RMSE Test,R2 Test
0,Regresión Lineal (OLS),0.773148,0.953421,0.976433,0.999973
1,Regresión Lasso,0.803091,1.034922,1.017311,0.999971
2,Regresión Rígida,0.772655,0.952406,0.975913,0.999973
3,Vecinos más cercanos (KNN),50.040133,4579.891556,67.674896,0.872031


### 1.6. Interpretación de resultados.
* Justifica cuál de los modelos utilizarías para ponerlo en producción

De acuerdo con los resultados que se pueden ver, podríamos decir que la mejor elección sería usar la **regresión rígida** y en segunda posición la **regresión lineal tradicional**

- Mejores métricas en Test: Es el modelo con el error más bajo de todos con un MAE de 0.7726 y un RMSE de 0.9759. Supera ligeramente a la regresión lineal tradicional (OLS).
- Ajuste casi perfecto: Tiene un $R^2$ de 0.9999, lo que significa que el modelo explica prácticamente toda la varianza de los datos.
- Robustez: Al usar penalización $L2$ (con $\alpha=0.1$), el modelo parece ser más estable frente al ruido que el OLS normal. Esto nos ayuda a evitar un posible sobreajuste.
- KNN descartado: El modelo KNN no funciona bien en este caso (MAE de 50.04).

---

## 2. Problema de Clasificación con generación de datos sintéticos y selección de características

En esta actividad se trabajará con generación de datos sintéticos para problemas de clasificación utilizando la función `make_classification` de la librería **scikit-learn**.

El objetivo es evaluar si distintos métodos de selección de características y modelos de clasificación son capaces de identificar las variables realmente relevantes dentro de un conjunto de datos generado artificialmente.

### 2.1 Importación de librerías

Importe las siguientes librerías:

In [44]:
import numpy as np
import pandas as pd

from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC

from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from sklearn.preprocessing import StandardScaler

### 2.2. Generación del conjunto de datos

2.2.1. Utilice `make_classification` para generar un dataset de clasificación binaria.

Debe definir explícitamente los siguientes parámetros:

- `n_samples`
- `n_features`
- `n_informative`
- `n_redundant`

El dataset debe tener:

- al menos **500 observaciones**
- al menos **10 características**

El objetivo es que existan variables:

- **informativas**
- **redundantes**
- **no informativas**

2.2.2. Explique brevemente los valores elegidos para cada parámetro.


### 2.3. División de los datos.

Divida el dataset en:

- **conjunto de entrenamiento**
- **conjunto de prueba**

El conjunto de entrenamiento se utilizará para:

- selección de características
- búsqueda de hiperparámetros
- entrenamiento de los modelos

El conjunto de prueba se utilizará **únicamente para evaluar el desempeño final de los modelos**.

### 2.4. Selección de caractarística

Aplique un método de selección de características para identificar cuáles variables parecen ser más relevantes para la predicción.

Puede utilizar el método **SelectKBest** de `sklearn.feature_selection`. Este método evalúa cada característica de forma individual mediante una medida estadística y selecciona las **k variables con mayor puntuación**, es decir, aquellas que tienen mayor relación con la variable objetivo.

Utilice como función de puntuación `f_classif`.

Compare las características seleccionadas con las definidas como **informativas** en `make_classification` y discuta si el método logra identificarlas correctamente.

## 2.5. Escalado de datos

Antes de entrenar los modelos, estandarice las variables utilizando **StandardScaler** de `sklearn.preprocessing`.

La estandarización consiste en transformar las variables para que tengan:

- media igual a 0
- desviación estándar igual a 1

Este paso es importante para modelos sensibles a la escala de los datos, como **regresión logística** y **SVM**. Los árboles de decisiones, son independientes del escalado.

El escalado debe realizarse de la siguiente manera:

1. Ajustar el `StandardScaler` **únicamente con el conjunto de entrenamiento**.
2. Aplicar esa transformación tanto al conjunto de entrenamiento como al conjunto de prueba.


### 2.6. Modelos e hiperparametrización

Trabaje con los siguientes modelos.

**Modelos obligatorios**

- Árbol de decisión (`DecisionTreeClassifier`)
- Regresión logística (`LogisticRegression`)

**Modelo bonus**

- Máquina de soporte vectorial (`SVC`)

Para cada modelo, realice una búsqueda de hiperparámetros utilizando **GridSearchCV** sobre el **conjunto de entrenamiento**.

Utilice las siguientes listas de hiperparámetros:

**Árbol de decisión**

- `max_depth`: [3, 5, 10, None]
- `min_samples_split`: [2, 5, 10]
- `criterion`: ["gini", "entropy"]

**Regresión logística**

- `C`: [0.01, 0.1, 1, 10]
- `solver`: ["lbfgs", "liblinear"]

**SVM (bonus)**

- `C`: [0.1, 1, 10]
- `kernel`: ["linear", "rbf"]
- `gamma`: ["scale", "auto"]

Utilice validación cruzada dentro de `GridSearchCV` para determinar la mejor combinación de hiperparámetros.


### 2.7. Evaluación y análisis

Para cada modelo:

- Del punto anterior, obtenga el modelo con **los mejores hiperparámetros**.
- Evalúe el modelo final en el **conjunto de prueba**.
- Realice este proceso en dos escenarios:
  - utilizando **todas las características**
  - utilizando **solo las características seleccionadas**
- Compare los resultados utilizando métricas de clasificación apropiada y discuta:
  - si la selección de características mejora el desempeño del modelo
  - si reduce la complejidad del modelo
  - si permite identificar correctamente las variables informativas generadas en el dataset
  - qué modelo obtiene mejor desempeño después del ajuste de hiperparámetros
